lib/network_monitor.c

```diff
@@ -562,59 +628,140 @@ static void handle_link(uint16_t type,
         }
 }
 
+/**
+ * @struct mptcpd_addr_info
+ *
+ * @brief Convenience structure to bundle address information.
+ */
+struct mptcpd_addr_info
+{
+        /// Network interface information.
+        struct mptcpd_interface *const interface;
+
+        /// Network address information.
+        struct sockaddr *const address;
+};
+
+/**
+ * @brief Notify new network address event subscriber.
+ *
+ * @param[in] data      Set of network event tracking callbacks.
+ * @param[in] user_data Network address information.
+ */
+static void notify_new_address(void *data, void *user_data)
+{
+        struct mptcpd_nm_ops    const *ops  = data;
+        struct mptcpd_addr_info const *info = user_data;
+
+        if (ops->new_address)
+                ops->new_address(info->interface, info->address);
+}
+
+/**
+ * @brief Notify new network address event subscriber.
+ *
+ * @param[in] data      Set of network event tracking callbacks.
+ * @param[in] user_data Network address information.
+ */
+static void notify_delete_address(void *data, void *user_data)
+{
+        struct mptcpd_nm_ops    const *ops  = data;
+        struct mptcpd_addr_info const *info = user_data;
+
+        if (ops->delete_address)
+                ops->delete_address(info->interface, info->address);
+}
+
 /**
  * @brief Register network address with network monitor.
  *
- * @param[in] addrs    List of tracked network addresses.
- * @param[in] rtm_addr New network address information.
+ * @param[in] nm        @c mptcpd_nm object that contains the list
+ *                      (queue) of network monitoring event
+ *                      subscribers.
+ * @param[in] interface @c mptcpd network interface information.
+ * @param[in] rtm_addr  New network address information.
  */
-static void insert_addr(struct l_queue *addrs,
+static void insert_addr(struct mptcpd_nm *nm,
+                        struct mptcpd_interface *interface,
                         struct mptcpd_rtm_addr const *rtm_addr)
 {
         struct sockaddr *const addr = mptcpd_sockaddr_create(rtm_addr);
 
         if (unlikely(addr == NULL)
-            || !l_queue_insert(addrs,
+            || !l_queue_insert(interface->addrs,
                                addr,
                                mptcpd_sockaddr_compare,
                                NULL)) {
                 mptcpd_sockaddr_destroy(addr);
 
                 l_error("Unable to track internet address information.");
         }
+
+        // Notify new network address event observers.
+        struct mptcpd_addr_info info = {
+                .interface = interface,
+                .address   = addr
+        };
+
+        l_queue_foreach(nm->ops, notify_new_address, &info);
```

Comment:
This causes the `new_address` op to be called when mptcpd explicitly populates the network address list in each `mptcpd_interface` at process start.  Such notifications should only occur later on mptcpd gets dynamically notified of a new address.  Move this `update_addr()`.  Use `update_link()` as a reference.